pkg/stream/connpool/msgconnpool/activeclient.go

```diff
@@ -0,0 +1,174 @@
+func (ac *activeClient) setHeartBeater(kp KeepAlive) {
+	if kp == nil {
+		return
+	}
+
+	kp.SaveHeartBeatFailCallback(func() {
+		ac.getConnData().Connection.Close(api.NoFlush, api.LocalClose)
+	})
+
+	ac.keepAlive = &keepAliveListener{
+		keepAlive: kp,
+		conn:      ac.getConnData().Connection,
+	}
+
+	ac.getConnData().Connection.AddConnectionEventListener(ac.keepAlive)
+}
+
+func (ac *activeClient) clearHeartBeater() {
+	// clear the previous keepAlive
+	if ac.keepAlive != nil && ac.keepAlive.keepAlive != nil {
+		ac.keepAlive.keepAlive.Stop()
+		ac.keepAlive = nil
+	}
+}
+
+// reconnect triggers connection to reconnect
+func (ac *activeClient) reconnect(event api.ConnectionEvent) {
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		return
+	}
+
+	if ac.connectTryTimes >= ac.pool.connTryTimes {
+		log.DefaultLogger.Warnf("[connpool] retry time exceed pool config %v", ac.pool.connTryTimes)
+		return
+	}
+
+	utils.GoWithRecover(func() {
+		time.Sleep(getBackOffDuration(ac.connectTryTimes))
+
+		ac.pool.connectingMux.Lock()
+		defer ac.pool.connectingMux.Unlock()
+
+		ac.initConnectionLocked(string(event))
+
+	}, func(r interface{}) {
+		log.DefaultLogger.Warnf("[connpool] panic while reconnect, %v, connData: %v, event : %v, host : %v",
+			r, ac.connData, event, ac.pool.Host().AddressString())
+	})
+
+}
+
+func (ac *activeClient) initConnectionLocked(initReason string) {
+	if atomic.LoadUint64(&ac.pool.destroyed) == 1 {
+		// if pool was exited, then stop
+		return
+	}
+
+	ac.connectTryTimes++
+
+	// build new conn
+	// must create this new conn, the same conn can only be connected once
+	createConnData := ac.pool.Host().CreateConnection(context.Background())
+	createConnData.Connection.AddConnectionEventListener(ac)
+
+	// connect the new connection
+	err := createConnData.Connection.Connect()
+
+	if err != nil {
+		if ac.getConnData() == nil {
+			// the first time
+			// atomic store, avoid partial write
+			ac.connData.Store(&createConnData)
+		}
+
+		log.DefaultLogger.Warnf("[connpool] connect failed %v times, host: %v, connData: %v, init reason: %v",
+			ac.connectTryTimes, ac.pool.Host().AddressString(), ac.connData, initReason)
+		return
+	}
+
+	// atomic store, avoid partial write
+	ac.connData.Store(&createConnData)
+
+	log.DefaultLogger.Infof("[connpool] connect succeed after %v tries, host: %v, connData: %v, init reason: %v",
+		ac.connectTryTimes, ac.pool.Host().AddressString(), ac.connData, initReason)
+
+	// init read filters and keep alive
+	if ac.pool.getReadFilterAndKeepalive != nil {
+		readFilterList, keepalive := ac.pool.getReadFilterAndKeepalive()
+
+		for _, rf := range readFilterList {
+			ac.getConnData().Connection.FilterManager().AddReadFilter(rf)
+		}
+
+		// set the new keepalive
+		ac.setHeartBeater(keepalive)
+	}
+
+	// clear retry times
+	ac.connectTryTimes = 0
+}
+
+func (ac *activeClient) OnEvent(event api.ConnectionEvent) {
+	switch {
+	case event.IsClose():
+		ac.clearHeartBeater()
```

Comment:
clearHeartBeater maybe called before setHeartBeater when connection instant break

Reply:
this break triggers a reconnection, and after acquiring the lock, the heartbeat will be set